3주차 미션 제출 (이선환)

glowing713 opened the pull request:
# 3주차 미션 제출 (이선환)

## 접근 방법

- **문서화(Github issues)**
개발에 앞서서 프로젝트를 설계하는 시간을 넉넉하게 가졌습니다. 2주차 과제에서는 개발해야할 작업을 따로 명시해두지 않다보니, 작업의 진행률이 어느정도인지 감을 잡기가 어려웠습니다. 따라서 전체적인 일정을 계획하고 문서화를 해보아야겠다고 생각하게 되었습니다.
프로젝트 코드가 위치한 곳에 기록을 남기는 것이 가장 보관에 용이하다고 생각하게 되어서, Github issue에 기록하고 이를 토대로 개발을 진행하게 되었습니다.

- **서버 API 통신을 가정한 상황**
  실서비스의 경우 데이터를 서버로부터 받아오는 과정이 존재하기 때문에, 제 프로젝트는 로컬 데이터를 사용하더라도 mount 상황 전에 데이터를 바인딩하도록 구현하였습니다. 약간의 딜레이를 위해 Promise 객체를 생성하여 사용하였습니다.

- **사용자 친화적 UI**
  어떤 식으로 아이템을 배치하는 것이 사용자에게 친화적일까 고민을 하게 되었습니다. 실서비스 중인 애플리케이션들의 HTML/CSS를 참고하며 프로젝트에 적용하였습니다.

- 개발을 진행하며 고민했던 내용과 해결한 이슈는 영상에 담았습니다. 다소 영상 길이가 길어서,, 죄송합니다,,🥲

## 특이사항

- 무한 스크롤링 구현
  지금 구현대로라면 예를 들어 렌더링하는 상품이 1000개라면, created 상태에 1000개를 불러오는 상황이 발생합니다. 사용자가 한번에 보는 상품의 개수는 6개에서 많아도 20개정도이므로, 약 2,30개씩 불러와서 상품을 보는 도중 계속해서 api 호출을 하고 렌더링하는 방법은 없을까 고민해보게 되었습니다. 그러던 중, infinite scroll을 알게 되었고, 적절한 라이브러리도 찾게 되었습니다,, 추후에 꼭 구현해보고 싶다는 생각이 들었습니다..!

## 제출 후 확인

* [x] 본인의 구현 결과를 `dev` 브랜치에 구현하고, `dev` 브랜치에서 `submission` 브랜치로 PR을 보내주세요.

CC: @externship-master  @lkaybob

glowing713 committed:
refactor: Refactor directories

glowing713 committed:
test: Check header container rendering

glowing713 committed:
feat: Add base-header component

glowing713 committed:
test: Check nav element rendering

glowing713 committed:
feat: Add base-nav component

glowing713 committed:
test: Check mall name is rendered in header

glowing713 committed:
feat: Add mall name paragraph in base-header

glowing713 committed:
test: Check notice button is rendered in base-header

glowing713 committed:
feat: Add notice nav button in base-header

glowing713 committed:
test: Check mall name anchors to main page(/)

glowing713 committed:
feat: Add anchor(href="/") to mall name

glowing713 committed:
style: Add style to header

glowing713 committed:
test: Check ul element rendering(buttons)

glowing713 committed:
feat: Add ul tag for buttons

glowing713 committed:
test: Check 5 buttons rendered in nav component

glowing713 committed:
feat: Add 5 buttons in BaseNav component

glowing713 committed:
style: Add style to BaseNav component

glowing713 committed:
fix: Fix height styling to include padding

glowing713 committed:
test: Check item lists container rendering

glowing713 committed:
feat: Add item lists container div

glowing713 committed:
test: Check thumbnail container rendering

glowing713 committed:
feat: Add thumbnail container

glowing713 committed:
test: Check content container rendering

glowing713 committed:
feat: Add content container

glowing713 committed:
test: Check item price rendering

glowing713 committed:
feat: Add price span and prop property

glowing713 committed:
test: Check discounted price is rendering on item sale

glowing713 committed:
feat: Add price calculating method in computed property

glowing713 committed:
test: Check discount rate is rendering if item is on sale

glowing713 committed:
feat: Add sale checking method in computed property

glowing713 committed:
chore: Add toLocaleString method to price

glowing713 committed:
test: Check market name, item name, sell count rendering

glowing713 committed:
feat: Add market name, item name, sell count span

glowing713 committed:
chore: Fixed typo

glowing713 committed:
test: Check sold count is formatted

glowing713 committed:
feat: Add sold count formatting

glowing713 committed:
chore: Add % to discount rate

glowing713 committed:
feat: Add item list mock data(js object)

glowing713 committed:
feat: Add item component and style

glowing713 committed:
style: Style item component

glowing713 committed:
chore: Fixed typo

glowing713 committed:
style: Add distance between itemlist and nav(padding-bottom)

glowing713 committed:
feat: Add mock server api process

glowing713 committed:
feat: Add scroll event

glowing713 committed:
chore: Fix typo

k12ms26 commented on mission/src/components/BaseComponent/BaseNav.vue:
```diff
@@ -0,0 +1,72 @@
+  name: 'BaseNav',
+  data() {
+    return {
+      buttonIcons: [
```
icon data는 따로 js 파일로 빼지 않으신 이유가 있으실까요??

glowing713 replied:
음.. 아이콘 데이터가 많거나 복잡하지는 않아서 컴포넌트 코드 내에 포함시키는게 좀 더 직관적이지 않을까..? 싶은 생각이었습니다!ㅎㅎ

k12ms26 reviewed (commented):
[좋은 점/잘한 점]
화면 크기에 따라 한줄에 dp되는 상품이 바뀌는 점에 놀랐습니다! bb
검색 버튼도 추가하신 센스가 돋보입니다

[의견제시]
코멘트 달아두었습니다 :)

daaeaad commented on mission/src/components/ItemList/Item.vue:
```diff
@@ -1,12 +1,188 @@
-  <div class="item-list-item"></div>
+  <div class="item-list-item">
+    <div data-test="thumbnail" class="thumbnail-container">
+      <img :src="thumbnail" alt="404 not found" />
```
alt 속성을 사용해서 이미지가 뜨지 않았을 때까지 대비 하셨네요!
저는 놓친 부분이라 다음부터는 저도 이런 세세한 것까지 신경써야겠다고 느꼈습니다!
다만 조금 아쉬운 부분이 있어서 alt속성에 들어갈 적절한 내용에 대한 의견을 남겨봅니다..!

제가 알기로 img 태그의 alt속성은 이미지가 뜨지 않았을 때나, 혹은 시각 장애인 사용자가 스크린리더를 사용할 때 사용 편의성을 높여주는 목적이므로 이미지에 대한 설명을 대체하는 내용이 들어가는 게 좋을것 같습니다!
따라서 404 메세지 보다는 해당 썸네일의 상품명이나 연결되는 링크를 남겨주시는 것이 좋을 것 같다고 생각했습니다!

예를 들면, '스웨터 이미지' 혹은 좀 더 자세하게 하면, '스웨터 썸네일 이미지 입니다. 클릭하면 상품의 상세페이지로 이동합니다.' 와 같은 텍스트가 적절할 것 같습니다 :)


**[ 참고자료 - 대체 텍스트 예시 ]**
- [https://brunch.co.kr/@snclab/2](url)

**[ 참고자료 - img alt속성에 대한 설명 ]**
- [http://tcpschool.com/html-tag-attrs/img-alt](url)


예전에 재직했던 회사에서 진행했던 프로젝트 중 웹접근성 테스트 통과 때문에 매우 고생했던 기억이 있어서요.
큰 문제는 아니지만 혹시 선환님께 나중에 도움이 될 수 있지 않을까 싶어 남겨봅니다 .. 🥲

glowing713 replied:
우와... 지금까지 img 태그를 사용하면서 대체 텍스트가 이렇게 중요한 기능을 한다고는 생각한 적이 없었는데.. 설령 이미지가 뜨지 않았거나, 스크린리더를 사용하더라도 전달하려는 정보에 오류가 없어야하겠군요..!! 가볍게 생각해서 놓치고 있었는데 사용자 입장에서는 정말 중요한 부분이라는걸 또 배워갑니다..!! 감사드려요😄

daaeaad commented on mission/src/components/ItemList/Item.vue:
```diff
@@ -1,12 +1,188 @@
+      </div>
+      <div class="sell-info">
+        <p v-if="sold" data-test="sell-count" class="sell-count">
+          {{ sold.toLocaleString() }}개 구매중
```
저도 테스트 과정에서 toLocaleString 때문에 자꾸 코드가 깨지는 원인을 끝내 파악하지 못해서 속상했는데..
v-if로 해결할 수 있었네요.. 저도 적용 해봐야겠습니다! 좋은 정보 감사합니다 😂

daaeaad reviewed (commented):
**[ 좋은점/잘한점 ]**
- nav에 들어갈 버튼을 v-for 디렉티브를 사용해서 반복되는 코드를 줄인 점이 배워갈 아이디어라고 생각합니다!
- 테스트 코드도 꼼꼼하게 작성하셔서 많이 참고하게 됩니다 :)
- 저도 toLocaleString 때문에 테스트 코드가 깨져서 속상했는데, 선환님 영상 보고 해결 방법을 얻은 것 같아요! 문제 원인 파악과 해결 능력이 뛰어나다고 생각했습니다.
- 항상 느끼는 거지만, 선환님은 코드도 그렇고 문서 내용 정리를 깔끔하게 잘 하시는 것 같아요! 
발표 영상도 늘 매끄럽게 잘 진행하시는 것 같아 0개 국어인 저는 많이 부럽습니다 .. ´, .̫ .`)

**[ 의견 ]**
- 코멘트 남겨두었습니다. 이번 주차도 고생 많으셨어요! 즐거운 설 연휴 보내세요 :)

2taesung commented:
[ 좋은점/잘한점 ]

- 너무 잘하시는 것 같아요! 앞으로 제시될 미션 주제들도 보이는 것 같습니다!(참고해야겠네요😏)
- 라이프 사이클을 이용해 스크롤 이벤트로 해더 히든 구현 하신 것 멋집니다...!!

ohyein00 commented:
- 발표를 너무 잘하시는 것 같습니다! 저도 localeString오류를 겪어 어찌 해결은 했었는데 명확히 알지 못해 찝찝한 상황이었는데 선환님의 명쾌한 설명을 듣고 나니 속이 뻥 뚤리는 것 같아요 🥰🥰
- 문서화를 정말 잘 하신 것 같습니다! 저는 잘 정리된 문서를 만드는 것에 늘 어려움을 겪는데 선환님을 본보기삼아 저도 다음 주차에서는 좀 더 문서화 하는 것에 노력을 기울여야겠다는 생각을 했습니다👍 감사합니다!
- nav를 v-for로 정리하신 부분이 인상 깊었습니다! 저도 참고해서 다음번에 적용해보도록 해야겠어요!
- 무한 스크롤링에 대한 아이디어가 좋다고 생각합니다! 저도 꼭 만들어보고 싶다는 생각을 했었는데 선환님께서도 추후 작업 계획이시라니 기대가 됩니다 😊😊

glowing713 committed:
chore: Add alt text to img tag
기존 img 태그에서 alt 속성에 이미지를 찾을 수 없다는 문구만 지정했었지만,
이미지가 렌더링되지 않더라도 정보의 전달에 오류가 없도록 product의 이름을
alt text로 지정하기로 변경

glowing713 committed:
refactor: Move rotate string project to Home component

lkaybob commented on mission/src/components/BaseComponent/BaseNav.vue:
```diff
@@ -0,0 +1,72 @@
+      buttonIcons: [
+        {
+          id: 'home',
+          html: '<i class="fas fa-home" />',
```
html을 직접 주는 것보다는 array syntax로 class를 넘기는 것이 더 좋아보입니다. 참고해보세요 (i tag는 template에)

https://v3.vuejs.org/guide/class-and-style.html#array-syntax

glowing713 replied:
넵! 감사합니다! 추천하신 방식이 좀 더 vue의 문법을 많이 사용하는 구문이라서 추천하시는걸로 이해하면 될까요..?

```vue
<template>
  <nav data-test="nav-container" class="nav-container">
    <ul>
      <li><i :class=["fas", "fa-home"] /></li>
      <li><i :class=["fas", "fa-search"] /></li>
      ...
    </ul>
  </nav>
</template>
```

그리고 이렇게 수정하면 될지 궁금합니다..!

lkaybob replied:
네 맞습니다.

glowing713 replied:
감사합니다~!

lkaybob commented on mission/tests/unit/ItemList/Item.spec.js:
```diff
@@ -1,10 +1,82 @@
+      const discountRate = wrapper.find('[data-test="discount-rate"]');
+
+      expect(discountRate.exists()).toBe(true);
+      expect(discountRate.text()).toEqual(`${String(10)}%`); // String 타입이라는 것을 명시하기 위해 형변환을 시킴
```
형변환은 필요없어 보입니다. "10%"라는 String만 맞으면 되는거라서 10이 string이어야한다는 것을 명시할 필요가 없다고 봅니다.

lkaybob commented on mission/tests/unit/ItemList/ItemList.spec.js:
```diff
@@ -1,10 +1,32 @@
+  it('renders item lists container', () => {
+    const itemLists = wrapper.find('[data-test="item-list-container"]');
+
+    expect(itemLists.exists()).toBe(true);
```
itemlist의 갯수에 따라 rendering되는 component 갯수를 검증하는 부분도 들어갔으면 좋아보입니다.

glowing713 replied:
넵 tc 추가해보겠습니다~!

lkaybob commented on mission/src/components/ItemList/Item.vue:
```diff
@@ -1,12 +1,188 @@
+      </div>
+      <div class="sell-info">
+        <p v-if="sold" data-test="sell-count" class="sell-count">
+          {{ sold.toLocaleString() }}개 구매중
```
inline expression은 지양하면서 이런 표현식은 computed property로 빼는 것이 더 좋아보입니다.

lkaybob commented on mission/src/views/Home.vue:
```diff
@@ -2,17 +2,38 @@
   <div class="home">
     <img alt="Vue logo" src="../assets/logo.png">
     <HelloWorld msg="Welcome to Your Vue.js App"/>
+    <base-input v-model="inputText"/>
```
이 파일은 왜 바뀐건가요?
* PR에는 해당 주차에 대한 변경점들만 담겨있었으면 합니다.

glowing713 replied:
pr을 닫고 복습과제를 진행해도 되는줄 모르고 커밋을 해서 pr에 반영이 되어버렸습니다.. 유의하겠습니다~!

lkaybob commented on mission/tests/unit/ItemList/Item.spec.js:
```diff
@@ -1,10 +1,82 @@
+  let wrapper;
 
+  beforeEach(() => {
+    wrapper = mount(ItemListItem);
```
Item.vue에 대한 전반적인 테스트케이스들이 테스트용 prop을 받고, 이들이 잘 바인딩이 되어 표시되는지 확인해야 합니다.
* 예) 상품 이름, 상품 가격, 상품 설명 등

glowing713 replied:
아..! 추가적인 질문이 있습니다 강사님! 그렇다면 prop을 받는 컴포넌트는 mount 시에 반드시 propdata를 포함시켜주는 것이 맞다고 이해하면 될까요..?

lkaybob reviewed (commented):
기능적인 요구사항을 만족하고, 애플리케이션 코드와 테스트 코드에서 다소 미흡한 점이 보입니다. Inline comment를 참고해주세요
